lib/assert.js

```diff
@@ -341,6 +358,197 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
   }
 };
 
+function isPlainObject(obj) {
+  if (typeof obj !== 'object' || obj === null) {
+    return false;
+  }
+  const proto = ObjectGetPrototypeOf(obj);
+  return proto === ObjectPrototype || proto === null || ObjectPrototypeToString(obj) === '[object Object]';
+}
+
+/**
+ * Compares two objects or values recursively to check if they are equal.
+ * @param {any} actual - The actual value to compare.
+ * @param {any} expected - The expected value to compare.
+ * @param {boolean} [loose=false] - Whether to use loose comparison (==) or strict comparison (===). Defaults to false.
+ * @param {Set} [comparedObjects=new Set()] - Set to track compared objects for handling circular references.
+ * @returns {boolean} - Returns `true` if the actual value matches the expected value, otherwise `false`.
+ * @example
+ * // Loose comparison (default)
+ * compareBranch({a: 1, b: 2, c: 3}, {a: 1, b: '2'}); // true
+ *
+ * // Strict comparison
+ * compareBranch({a: 1, b: 2, c: 3}, {a: 1, b: 2}, true); // true
+ */
+function compareBranch(
+  actual,
+  expected,
+  loose = false,
+  comparedObjects = new SafeSet(),
+) {
+  // Check for Map object equality
+  if (isMap(actual) && isMap(expected)) {
+    if (actual.size !== expected.size) {
+      return false;
+    }
+    const safeIterator = FunctionPrototypeCall(SafeMap.prototype[SymbolIterator], actual);
+
+    for (const { 0: key, 1: val } of safeIterator) {
+      if (!MapPrototypeHas(expected, key)) {
+        return false;
+      }
+      if (!compareBranch(val, MapPrototypeGet(expected, key), loose, comparedObjects)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  // Check for Set object equality
+  if (isSet(actual) && isSet(expected)) {
+    if (actual.size !== expected.size) {
```

Comment:
The equality seems weird, wouldn't we want to support `actual` with more elements than `expected`?
```suggestion
    // TODO(aduh95): switch to `SetPrototypeIsSubsetOf` when it's available
    if (actual.size > expected.size) {
```
BTW (and sorry it's very late in the conversation, I didn't think of it earlier), `assert.isSubsetOf` might have been a better name to reflect to the built-in `Set` method (https://tc39.es/ecma262/#sec-set.prototype.issubsetof)

Reply:
Good catch, it was reminiscence from OP which I didn't notice